source/slang/slang-check-conversion.cpp

```diff
@@ -388,88 +429,246 @@ namespace Slang
+                        {
+                            coercedArgs.add(coercedArg);
+                        }
+                    }
 
-                    if (coercedArg)
+                    // We will go through the fields in order and try to match them
+                    // up with initializer arguments.
+                    //
+                    for (auto fieldDeclRef : getMembersOfType<VarDecl>(m_astBuilder, toStructDeclRef, MemberFilterStyle::Instance))
                     {
-                        coercedArgs.add(coercedArg);
+                        Expr* coercedArg = nullptr;
+                        bool argResult = _readValueFromInitializerList<UseLegacyLogic>(
+                            getType(m_astBuilder, fieldDeclRef),
+                            outToExpr ? &coercedArg : nullptr,
+                            fromInitializerListExpr,
+                            ioArgIndex);
+
+                        // No point in trying further if any argument fails
+                        if (!argResult)
+                            return false;
+
+                        if (coercedArg)
+                        {
+                            coercedArgs.add(coercedArg);
+                        }
                     }
                 }
-
-                // We will go through the fields in order and try to match them
-                // up with initializer arguments.
-                //
-                for(auto fieldDeclRef : getMembersOfType<VarDecl>(m_astBuilder, toStructDeclRef, MemberFilterStyle::Instance))
+            }
+            else
+            {
+                auto toTypeDeclRef = toDeclRefType->getDeclRef();
+                if (auto toStructDeclRef = toTypeDeclRef.as<StructDecl>())
                 {
-                    Expr* coercedArg = nullptr;
-                    bool argResult = _readValueFromInitializerList(
-                        getType(m_astBuilder, fieldDeclRef),
-                        outToExpr ? &coercedArg : nullptr,
-                        fromInitializerListExpr,
-                        ioArgIndex);
+                    // We will try to coerce the initializer list into a struct following these steps:
+                    // 1. Ensure our `StructDecl` has declaratations created for all auto-generated constructors
+                    // 2. We have a `{}` (0 arguments), try to coerce
+                    // 3. We have a `{arg1, arg2...}`, try to coerce
 
-                    // No point in trying further if any argument fails
-                    if(!argResult)
-                        return false;
+                    auto toStructDecl = toStructDeclRef.getDecl();
+                    // 1. Ensure our `StructDecl` has declaratations created for all auto-generated constructors
+                    ensureDecl(toStructDecl, DeclCheckState::DefaultConstructorReadyForUse);
 
-                    if( coercedArg )
+                    // 2. We have a `{}` (0 arguments), try to coerce
```

Comment:
I think we should delete all these logic here from 535 till 671, and reuse the existing `ResolveInvoke` logic to resolve the constructor call.

Once we reach here, we are about to turn initializer list `{...}` into a ctor call that is `StructType(...)`.
This can be done by:
1. create a AST node that is `InvokeExpr(calleeExpr, coercedArgs)`, where:
    `calleeExpr` is a `VarExpr(declRef = toStructDeclRef, type = TypeType(toType), checked = true)`.
2. call `ResolveInvoke` on the invokeExpr, and that is the `*outToExpr` we need to return.

This should handle all the generic case or the variadic case just fine using the same logic in ResolveInvoke.

Reply:
Create a sub-task to address this issue: https://github.com/shader-slang/slang/issues/5134.
Note, I won't submit a new PR for this. It just that I found out the original task took much longer than we estimated (estimation is just 1 day! And I already spent 3 days). And it could take longer, so split the task smaller so that I will have a clear goal what I'm going to resolve.